Enterprise-CMCS/macpro-mdct-qmr
# Upgrade cypress and prep year update

### Description
- Overhaul QMR cypress to test against 2023
  - Upgrade cypress to v12 for faster test running and upgraded features
  - cleanup/upgrade dependencies
    - remove `xpath`
    - swap imported `attachFile` for native `selectFile`
  - update config files

This PR sets up the above. It will have multiple stacked PRs to finish the changes for each spec file. The plan is to merge all of those stacked PRs into this branch, then merge this into master. Please review this first, then stacked PRs.

Stacked PRs
https://github.com/Enterprise-CMCS/macpro-mdct-qmr/pull/1944
https://github.com/Enterprise-CMCS/macpro-mdct-qmr/pull/1946
https://github.com/Enterprise-CMCS/macpro-mdct-qmr/pull/1947

Merged
https://github.com/Enterprise-CMCS/macpro-mdct-qmr/pull/1929
https://github.com/Enterprise-CMCS/macpro-mdct-qmr/pull/1930
https://github.com/Enterprise-CMCS/macpro-mdct-qmr/pull/1932
https://github.com/Enterprise-CMCS/macpro-mdct-qmr/pull/1933
https://github.com/Enterprise-CMCS/macpro-mdct-qmr/pull/1934
https://github.com/Enterprise-CMCS/macpro-mdct-qmr/pull/1935
https://github.com/Enterprise-CMCS/macpro-mdct-qmr/pull/1936
https://github.com/Enterprise-CMCS/macpro-mdct-qmr/pull/1937
https://github.com/Enterprise-CMCS/macpro-mdct-qmr/pull/1938
https://github.com/Enterprise-CMCS/macpro-mdct-qmr/pull/1939
https://github.com/Enterprise-CMCS/macpro-mdct-qmr/pull/1940
https://github.com/Enterprise-CMCS/macpro-mdct-qmr/pull/1941
https://github.com/Enterprise-CMCS/macpro-mdct-qmr/pull/1942
https://github.com/Enterprise-CMCS/macpro-mdct-qmr/pull/1943


See notes below on recurring changes that will help simplify your review

I also noticed many of these files have repeat tests so there is plenty of opportunity to improve these and make it much easier to do something like this again for future year updates 👀 

### Related ticket(s)
MDCT-2554

---
### How to test
Cypress tests pass


### Important updates
Be sure to re-yarn your root and tests/cypress folders!


### Recurring changes
You will see a new constant import in every file as that is how we are controlling which year to test (and it means in the future we can just update it in one location!)
```
import { testingYear } from "../../../../support/constants";
...
cy.selectYear(testingYear);
```

We modified the ids for certain fields to not be linked to their title (nor inaccurate) so you'll see many updates that just swaps a value in the `data-cy` path for a random string that is the new id
```
- PerformanceMeasure.rates.singleCategory.0.rate
+ PerformanceMeasure.rates.xacc8a.0.rate
```

Removed `force: true` from the clicks because that's just weird
```
- .click({ force: true });
+ .click();
```

xpath is deprecated so we are changing it for the easier to read and already frequently used `get`
```
- cy.xpath("(//input[@label='Month'])[1]").type("10");
+ cy.get('[data-cy="DateRange.startDate-month"').type("10");
```


`attachFile` was an imported function that has since been integrated in cypress natively as `selectFile` (note: these fail locally because file uploads fail locally. I also opted for the direct upload rather than drag-and-drop. let me know if that's too much of a stretch)
```
- cy.xpath(browseBTN).attachFile(filePath + "test3.docx", {
-     subjectType: "drag-n-drop",
- });
+ cy.get('[data-testid="upload-component"]')
+     .invoke("show")
+     .selectFile(`${filePath}test3.docx`);
```
